File: SSUMobile/Modules/Calendar/Views/SSUCalendarViewController.m
//
//  SSUCalendarMonthViewController.m
//  SSUMobile
//
//  Created by Eric Amorde on 9/28/14.
//  Copyright (c) 2014 Sonoma State University Department of Computer Science. All rights reserved.
//

#import "SSUCalendarViewController.h"
#import "SSUSelectionController.h"
#import "SSUConfiguration.h"
#import "SSUMobile-Swift.h"

#import <Masonry/Masonry.h>

static NSString * const SSUEventDetailSegue = @"event";
static NSString * const EventCellIdentifier = @"EventCell";

static CGFloat TABLE_HEADER_HEIGHT = 25.0;
static CGFloat CELL_ROW_HEIGHT = 50;

// Hack to get access to date property
@interface PDTSimpleCalendarViewController ()

- (NSDate *) dateForCellAtIndexPath:(NSIndexPath *)indexPath;
- (BOOL)isEnabledDate:(NSDate *)date;

@end

@interface SSUCalendarViewController () <PDTSimpleCalendarViewDelegate,UITableViewDataSource,UITableViewDelegate, SSUSelectionDelegate>

@property (nonatomic, strong) id selectedEvent;
@property (nonatomic) NSArray * unfilteredEvents;
@property (nonatomic) NSArray * events;
@property (nonatomic) NSArray * selectedEvents;
@property (nonatomic) NSPredicate * predicate;

@end

@implementation SSUCalendarViewController

- (void) loadEvents
{
    NSFetchRequest * fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"SSUEvent"];
    fetchRequest.sortDescriptors = @[[NSSortDescriptor sortDescriptorWithKey:@"startDate" ascending:YES]];
    NSManagedObjectContext * context = [[SSUCalendarModule sharedInstance] backgroundContext];
    [context performBlock:^{
        self.unfilteredEvents = [context executeFetchRequest:fetchRequest error:nil];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self reloadEventTableView];
            [self.collectionView reloadData];
        });
    }];
}

- (void) viewDidLoad {
    // We have to do these two lines before [super viewDidLoad] for it to show up :/
    self.weekdayHeaderEnabled = YES;
    [self applyStyles];
    [super viewDidLoad];
    
    self.delegate = self;
    self.navigationItem.title = @"Seawolf Calendar";
    
    self.tableView = [[UITableView alloc] initWithFrame:self.view.frame style:UITableViewStylePlain];
    [self.view addSubview:self.tableView];
    self.tableView.dataSource = self;
    self.tableView.delegate = self;
    self.tableView.hidden = YES;
    UINib * cellNib = [UINib nibWithNibName:NSStringFromClass([SSUCalendarEventCell class]) bundle:nil];
    [self.tableView registerNib:cellNib forCellReuseIdentifier:EventCellIdentifier];
    [self.tableView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.height.equalTo(self.view.mas_height).multipliedBy(1/3.0);
        make.width.equalTo(self.view.mas_width);
        make.bottom.equalTo(self.view.mas_bottom);
    }];
    
    UIBarButtonItem * filterButton = [[UIBarButtonItem alloc] initWithTitle:@"Filter" style:UIBarButtonItemStyleDone target:self action:@selector(filterButtonPressed:)];
    self.navigationItem.rightBarButtonItem = filterButton;
    
    [self loadEvents];
    self.selectedDate = [NSDate date];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    if (self.selectedDate == nil) {
        self.selectedDate = [NSDate date];
    }
    
    const NSTimeInterval lastUpdate = [[[SSUConfiguration sharedInstance] calendarLastUpdate] timeIntervalSinceNow];
    const NSTimeInterval updateInterval = -1 * 60 * 5;
    if (lastUpdate <= updateInterval) {
        [[SSUCalendarModule sharedInstance] updateData:^{
            [self loadEvents];
        }];
    }
}

- (void) applyStyles {
    [[PDTSimpleCalendarViewCell appearance] setCircleTodayColor:[UIColor blackColor]];
    [[PDTSimpleCalendarViewCell appearance] setCircleSelectedColor:SSU_BLUE_COLOR];
    
    [[PDTSimpleCalendarViewHeader appearance] setTextColor:SSU_BLUE_COLOR];
    [[PDTSimpleCalendarViewHeader appearance] setSeparatorColor:[UIColor clearColor]];

    [[PDTSimpleCalendarViewWeekdayHeader appearance] setHeaderBackgroundColor:SSU_BLUE_COLOR];
    [[PDTSimpleCalendarViewWeekdayHeader appearance] setTextColor:[UIColor whiteColor]];
}

- (BOOL)tableViewVisible {
    return !self.tableView.hidden;
}

- (void) reloadEventTableView {
    self.selectedEvents = [self eventsForDate:self.selectedDate];
    [self.tableView reloadData];
    self.tableView.hidden = self.selectedEvents.count == 0;
}

#pragma mark - Filtering

- (void) setPredicate:(NSPredicate *)predicate {
    _predicate = predicate;
    [self.collectionView reloadData];
    [self reloadEventTableView];
}

- (NSArray *) events {
    if (self.predicate) {
        return [self.unfilteredEvents filteredArrayUsingPredicate:self.predicate];
    }
    else {
        return self.unfilteredEvents;
    }
}

#pragma mark - Helper

- (NSDateComponents *) nonTimeComponents:(NSDate *)date {
    static NSCalendarUnit options = NSCalendarUnitDay | NSCalendarUnitMonth | NSCalendarUnitYear;
    NSDateComponents * components = [[NSCalendar currentCalendar] components:options fromDate:date];
    return components;
}

// Returns a date that is the same calendar date but with a time of 00:00:00
- (NSDate *) startOfDay:(NSDate *)date {
    NSDateComponents * components = [self nonTimeComponents:date];
    components.hour = 0;
    components.minute = 0;
    components.second = 0;

    return [[NSCalendar currentCalendar] dateFromComponents:components];;
}

// Returns a date that is the same calendar date but with a time of 23:59:59
- (NSDate *) endOfDay:(NSDate *)date {
    NSDateComponents * components = [self nonTimeComponents:date];
    components.hour = 23;
    components.minute = 59;
    components.second = 59;
    
    return [[NSCalendar currentCalendar] dateFromComponents:components];
}

// Returns an NSArray of event objects whos `startDate` is on the same day as date
- (NSArray *) eventsForDate:(NSDate *)date {
    if (date == nil) {
        return nil;
    }
    NSDate * start = [self startOfDay:date];
    NSDate * end = [self endOfDay:date];
    NSPredicate * predicate = [NSPredicate predicateWithFormat:@"startDate BETWEEN %@", @[start, end]];
    return [self.events filteredArrayUsingPredicate:predicate];
}

#pragma mark - PDTSimpleCalendarViewControllerDelegate

- (void) simpleCalendarViewController:(PDTSimpleCalendarViewController *)controller didSelectDate:(NSDate *)date {
    self.selectedEvents = [self eventsForDate:date];
    [self reloadEventTableView];
}

#pragma mark - UICollectionView

- (BOOL) indexPathIsLastWeek:(NSIndexPath *)indexPath {
    const NSInteger dayCount = [self.collectionView numberOfItemsInSection:indexPath.section] - 1;
    return (dayCount - indexPath.item) < 7;
}

- (BOOL) indexPathIsFirstWeek:(NSIndexPath *)indexPath {
    return indexPath.item < 7;
}

- (UICollectionViewCell *) collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    static const NSInteger labelTag = 1077;
    static const NSInteger separatorTag = 1099;
    PDTSimpleCalendarViewCell * cell = (id)[super collectionView:collectionView cellForItemAtIndexPath:indexPath];
    
    // PDTSimpleCalendar makes it somewhat difficult to use a cell subclass. So instead of making that work,
    // we will use tags to manually add subviews. Not nearly as clean, but it's better than trying to mess with
    // the inner workings of PDTSimpleCalendar. Maybe that will change in the future.
    
    NSDate * date = [self dateForCellAtIndexPath:indexPath];
    const NSUInteger count = [[self eventsForDate:date] count];
    UILabel * label = (UILabel *)[cell.contentView viewWithTag:labelTag];
    if (label == nil) {
        label = [[UILabel alloc] init];
        label.font = [UIFont preferredFontForTextStyle:UIFontTextStyleFootnote];
        label.textColor = [UIColor lightGrayColor];
        label.tag = labelTag;
        label.minimumScaleFactor = .50;
        label.adjustsFontSizeToFitWidth = YES;
        label.textAlignment = NSTextAlignmentCenter;
        [cell.contentView addSubview:label];
        [label mas_makeConstraints:^(MASConstraintMaker *make) {
            const CGFloat margin = 1.0;
            UILabel * dayLabel = [cell valueForKey:@"dayLabel"];
            make.top.equalTo(label.superview.mas_top);
            make.right.equalTo(label.superview.mas_right).with.offset(-margin);
            make.bottom.greaterThanOrEqualTo(dayLabel.mas_top).with.offset(-margin);
            make.left.greaterThanOrEqualTo(dayLabel.mas_right).with.offset(margin);
        }];
        [cell.contentView bringSubviewToFront:label];
    }
    NSString * eventText = [NSString stringWithFormat:@"%lu", (unsigned long)count];
    label.text = (count > 0) ? eventText : @"";
    
    // If we are on the overflow days (ex. before first day of month or after last day of month)
    // then we shouldn't show the event label
    label.hidden = [cell valueForKey:@"date"] == nil; // If only we actually had access to this..
    
    SSUCollectionCellSeparatorView * separatorView = (id)[cell viewWithTag:separatorTag];
    if (separatorView == nil) {
        separatorView = [[SSUCollectionCellSeparatorView alloc] initWithFrame:cell.bounds];
        [cell addSubview:separatorView];
        [cell sendSubviewToBack:separatorView];
        separatorView.separatorColor = [UIColor lightGrayColor];
    }
    
    SSUCellSeparator separatorSides = SSUCellSeparatorTop;
    if ([self indexPathIsLastWeek:indexPath]) {
        separatorSides |= SSUCellSeparatorBottom;
    }
    if (!(indexPath.item % 7 == 0)) {
        // Don't draw a border on the first day if we don't need it
        separatorSides |= SSUCellSeparatorLeft;
    }
    separatorView.separator = separatorSides;
    
    return cell;
}

#pragma mark - UITableView

- (void) tableView:(UITableView *)tableView willDisplayHeaderView:(UIView *)view forSection:(NSInteger)section {
    UITableViewHeaderFooterView * header = (UITableViewHeaderFooterView *)view;
    header.tintColor = [UIColor whiteColor];
    header.contentView.backgroundColor = SSU_BLUE_COLOR;
    header.textLabel.textColor = [UIColor whiteColor];
}

- (CGFloat) tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    return TABLE_HEADER_HEIGHT;
}

- (NSString *) tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    NSString * formattedDate = [NSDateFormatter localizedStringFromDate:self.selectedDate
                                                              dateStyle:NSDateFormatterShortStyle
                                                              timeStyle:NSDateFormatterNoStyle];
    
    return [NSString stringWithFormat:@"%lu events on %@",(unsigned long)self.selectedEvents.count,formattedDate];;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.selectedEvents.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    SSUCalendarEventCell *eventCell = [tableView dequeueReusableCellWithIdentifier:EventCellIdentifier forIndexPath:indexPath];
    SSUEvent *event = self.selectedEvents[indexPath.row];
    eventCell.event = event;
    
    return eventCell;
}

- (void) tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    self.selectedEvent = self.selectedEvents[indexPath.row];
    //[self performSegueWithIdentifier:SSUEventDetailSegue sender:self];
    
    UIStoryboard * storyboard = self.storyboard;
    SSUCalendarEventDetailController * detailViewController = [storyboard instantiateViewControllerWithIdentifier:NSStringFromClass([SSUCalendarEventDetailController class])];
    detailViewController.event = self.selectedEvent;
    [self.navigationController pushViewController:detailViewController animated:YES];
}

- (CGFloat) tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return CELL_ROW_HEIGHT;
}

#pragma mark - SSUSelectionDelegate

- (void) userDidSelectItem:(id)item
               atIndexPath:(NSIndexPath *)indexPath
            fromController:(SSUSelectionController *)controller {
    if ([item isEqualToString:@"All Categories"]) {
        self.predicate = nil;
    }
    else {
        NSString * categoryName = item;
        self.predicate = [NSPredicate predicateWithFormat:@"category = %@", categoryName];
    }
    
    [self.navigationController popViewControllerAnimated:YES];
}

- (void) selectionControllerDismissed:(SSUSelectionController *)controller {
    [self.navigationController popViewControllerAnimated:YES];
}

#pragma mark - Buttons

- (NSArray *) allCategories {
    NSFetchRequest * request = [NSFetchRequest fetchRequestWithEntityName:@"SSUEvent"];
    NSString * categoryKeyName = @"category";
    request.propertiesToFetch = @[categoryKeyName];
    request.returnsDistinctResults = YES;
    request.resultType = NSDictionaryResultType;
    
    NSArray * objects = [[SSUCalendarModule sharedInstance].context executeFetchRequest:request error:nil];
    NSMutableArray * categories = [NSMutableArray arrayWithCapacity:objects.count];
    for (NSDictionary * dict in objects) {
        if (dict[categoryKeyName] != nil) {
            [categories addObject:dict[categoryKeyName]];
        }
    }
    return [categories sortedArrayUsingSelector:@selector(localizedCaseInsensitiveCompare:)];
}

- (void) filterButtonPressed:(UIBarButtonItem *)button {
    NSArray * categories = [@[@"All Categories"] arrayByAddingObjectsFromArray:[self allCategories]];
    SSUSelectionController * selectionController = [[SSUSelectionController alloc] initWithItems:categories];
    selectionController.delegate = self;
    if (self.predicate == nil) {
        selectionController.defaultIndex = [NSIndexPath indexPathForRow:0 inSection:0];
    }
    else {
        NSString * current = [self.events[0] category];
        NSInteger row = (NSInteger)[categories indexOfObject:current];
        selectionController.defaultIndex = [NSIndexPath indexPathForRow:row inSection:0];
    }
    [self.navigationController pushViewController:selectionController animated:YES];
}

#pragma mark - Segue

- (void) prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqualToString:SSUEventDetailSegue])
    {
        SSUCalendarEventDetailController * detailController = segue.destinationViewController;
        detailController.event = self.selectedEvent;
    }
}

@end

File: SSUMobile/Modules/Calendar/Views/SSUCalendarViewController.swift
//
//  SSUCalendarViewController.swift
//  SSUMobile
//
//  Created by Eric Amorde on 4/4/17.
//  Copyright © 2017 Sonoma State University Department of Computer Science. All rights reserved.
//

import Foundation
import SnapKit

class SSUCalendarViewController: UIViewController, FSCalendarDataSource, FSCalendarDelegate, UITableViewDataSource, UITableViewDelegate, SSUSelectionDelegate {
    
    private struct Identifiers {
        static let eventCell = "EventCell"
        static let calendarCell = "calendarCell"
        static let eventSegue = "event"
    }
    
    let tableView = UITableView(frame: .zero, style: .plain)
    let calendarView = FSCalendar(frame: .zero)
    let selectEventLabel = UILabel()
    
    var context: NSManagedObjectContext = SSUCalendarModule.instance.context
    
    private var selectedEvent: SSUEvent?
    
    private var unfilteredEvents: [SSUEvent] = []
    private var selectedEvents: [SSUEvent] = []
    private var events: [SSUEvent] {
        if let filter = predicate {
            return unfilteredEvents.filter({ (event) -> Bool in
                filter.evaluate(with: event)
            })
        } else {
            return unfilteredEvents
        }
    }
    
    private var predicate: NSPredicate? {
        didSet {
            calendarView.reloadData()
            reloadEventTableView()
        }
    }
    
    private let tableViewHeaderHeight: CGFloat = 25.0
    
    override func viewDidLoad() {
        super.viewDidLoad()
        title = "Seawolf Calendar"
        
        applyStyles()
        setupViews()
        
        let filterButton = UIBarButtonItem(title: "Filter", style: .done, target: self, action: #selector(filterButtonPressed(button:)))
        navigationItem.rightBarButtonItem = filterButton
        
        calendarView.select(Date())
        calendarView.reloadData()
        loadEvents()
        reloadEventTableView()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        if calendarView.selectedDate == nil {
            calendarView.select(Date())
        }
        
        if let lastUpdate = SSUConfiguration.sharedInstance().calendarLastUpdate {
            let interval: TimeInterval = -1 * 60 * 5
            if lastUpdate.timeIntervalSinceNow <= interval {
                refresh()
            }
        } else {
            refresh()
        }
    }
    
    private func setupViews() {
        view.addSubview(calendarView)
        calendarView.snp.makeConstraints { (make) in
            make.top.equalToSuperview()
            make.right.equalToSuperview()
            make.left.equalToSuperview()
            make.height.equalTo(view.frame.height / 2.0)
        }
        calendarView.delegate = self
        calendarView.dataSource = self
        calendarView.scrollEnabled = true
        calendarView.scrollDirection = .vertical
        calendarView.scope = .month
        calendarView.backgroundColor = UIColor.white
        calendarView.register(SSUCalendarCollectionCell.self, forCellReuseIdentifier: Identifiers.calendarCell)
        
        view.addSubview(tableView)
        tableView.dataSource = self
        tableView.delegate = self
        tableView.estimatedRowHeight = 50
        tableView.register(SSUCalendarEventCell.self, forCellReuseIdentifier: Identifiers.eventCell)
        tableView.snp.makeConstraints { (make) in
            make.height.equalTo(0)
            make.left.equalToSuperview()
            make.right.equalToSuperview()
            make.bottom.equalToSuperview()
        }
        
        view.addSubview(selectEventLabel)
        view.sendSubview(toBack: selectEventLabel)
        selectEventLabel.snp.makeConstraints { (make) in
            make.top.equalTo(calendarView.snp.bottom)
            make.left.equalToSuperview()
            make.right.equalToSuperview()
            make.bottom.equalToSuperview()
        }
        selectEventLabel.textAlignment = .center
        if #available(iOS 9.0, *) {
            selectEventLabel.font = UIFont.preferredFont(forTextStyle: .title1)
        } else {
            selectEventLabel.font = UIFont.boldSystemFont(ofSize: 28.0)
        }
        selectEventLabel.backgroundColor = UIColor.groupTableViewBackground
        selectEventLabel.textColor = UIColor.ssuBlue
        selectEventLabel.text = "No events found"
    }
    
    private func refresh() {
        SSUCalendarModule.instance.updateData({
            self.loadEvents()
        })
    }
    
    private func loadEvents() {
        let fetchRequest: NSFetchRequest<SSUEvent> = SSUEvent.fetchRequest()
        fetchRequest.sortDescriptors = [
            NSSortDescriptor(key: #keyPath(SSUEvent.startDate), ascending: true)
        ]
        do {
            unfilteredEvents = try context.fetch(fetchRequest)
        } catch {
            SSULogging.logError("Error fetching events: \(error)")
            unfilteredEvents = []
        }
        
        DispatchQueue.main.async {
            self.reloadEventTableView()
            self.calendarView.reloadData()
        }
    }
    
    private func applyStyles() {
        calendarView.appearance.todayColor = .black
        calendarView.appearance.selectionColor = .ssuBlue
        calendarView.appearance.headerTitleColor = .white
        calendarView.appearance.weekdayTextColor = .white
        
        calendarView.calendarHeaderView.backgroundColor = .ssuBlue
        calendarView.calendarWeekdayView.backgroundColor = .ssuBlue
    }
    
    private func reloadEventTableView() {
        selectedEvents = eventsForDate(calendarView.selectedDate)
        tableView.reloadData()
        tableView.snp.updateConstraints { (make) in
            if selectedEvents.count > 0 {
                make.height.equalTo(view.frame.height - calendarView.frame.height)
            } else {
                make.height.equalToSuperview().multipliedBy(0.0)
            }
        }
    }
    
    private func eventsForDate(_ date: Date?) -> [SSUEvent] {
        guard let date = date else {
            return []
        }
        
        return self.events.filter({ (event) -> Bool in
            if let startDate = event.startDate {
                return Calendar.current.isDate(startDate, inSameDayAs: date)
            }
            return false
        })
    }
    
    // MARK: - FSCalendarDataSource
    
    func calendar(_ calendar: FSCalendar, cellFor date: Date, at position: FSCalendarMonthPosition) -> FSCalendarCell {
        let cell = calendar.dequeueReusableCell(withIdentifier: Identifiers.calendarCell, for: date, at: position)
        if let calendarCell = cell as? SSUCalendarCollectionCell {
            calendarCell.eventCount = eventsForDate(date).count
            calendarCell.separator = [.top, .bottom, .right, .left]
        }
        return cell
    }
    
    // MARK: - FSCalendarDelegate
    
    func calendar(_ calendar: FSCalendar, didSelect date: Date, at monthPosition: FSCalendarMonthPosition) {
        reloadEventTableView()
    }
    
    func calendar(_ calendar: FSCalendar, boundingRectWillChange bounds: CGRect, animated: Bool) {
        calendarView.snp.updateConstraints { (make) in
            make.height.equalTo(bounds.height)
        }
        
        reloadEventTableView()
    }
    
    // MARK: - UITableViewDataSource
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return selectedEvents.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell = tableView.dequeueReusableCell(withIdentifier: Identifiers.eventCell, for: indexPath) as? SSUCalendarEventCell else {
            SSULogging.logError("Unable to deqeue correct event cell type")
            return UITableViewCell()
        }
        
        cell.event = selectedEvents[indexPath.row]
        
        return cell
    }
    
    // MARK: - UITableViewDelegate
    
    func tableView(_ tableView: UITableView, willDisplayHeaderView view: UIView, forSection section: Int) {
        if let headerView = view as? UITableViewHeaderFooterView {
            headerView.contentView.backgroundColor = .ssuBlue
            headerView.tintColor = .white
            headerView.textLabel?.textColor = .white
        }
    }
    
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return tableViewHeaderHeight
    }
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        if let selectedDate = calendarView.selectedDate {
            let formattedDate = DateFormatter.localizedString(from: selectedDate, dateStyle: .short, timeStyle: .none)
            return "\(selectedEvents.count) events on \(formattedDate)"
        }
        return nil
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        selectedEvent = selectedEvents[indexPath.row]
        performSegue(withIdentifier: Identifiers.eventSegue, sender: self)
    }
    
    // MARK: - SSUSelectionDelegate
    
    func userDidSelectItem(_ item: Any!, at indexPath: IndexPath!, from controller: SSUSelectionController!) {
        if indexPath.item == 0 && indexPath.section == 0 {
            predicate = nil
        } else if let category = item as? String {
            predicate = NSPredicate(format: "%K = %@", argumentArray: [#keyPath(SSUEvent.category), category])
        }
        navigationController?.popViewController(animated: true)
    }
    
    func selectionControllerDismissed(_ controller: SSUSelectionController!) {
        navigationController?.popViewController(animated: true)
    }
    
    // MARK: - Buttons
    
    private func allCategories() -> [String] {
        let request: NSFetchRequest<SSUEvent> = SSUEvent.fetchRequest()
        request.propertiesToFetch = [#keyPath(SSUEvent.category)]
        
        let results = (try? context.fetch(request)) ?? []
        let categories = results.flatMap({ $0.category })
        return Array(Set(categories)).sorted()
    }
    
    func filterButtonPressed(button: UIBarButtonItem) {
        var categories = allCategories()
        categories.insert("All Categories", at: 0)
        
        let controller = SSUSelectionController(items: categories)!
        controller.delegate = self
        
        if predicate != nil, let currentCategory = events.first?.category, let row = categories.index(of: currentCategory) {
            controller.defaultIndex = IndexPath(row: row, section: 0)
        } else {
            controller.defaultIndex = IndexPath(row: 0, section: 0)
        }
        navigationController?.pushViewController(controller, animated: true)
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if let destination = segue.destination as? SSUCalendarEventDetailController {
            destination.event = selectedEvent
        }
    }
}
